### DISCUSSION TITLE
GPL and placement density

### DISCUSSION THREAD

**Discussion by lustefan97** [The-OpenROAD-Project/OpenROAD#4257] [Q&A]
Hello,

I have a question about the target placement density in the GPL tool. 

I don't really get what it does quantitatively speaking. I get that the closer target_density is close to 1, the tighter the placement will be, but it doesn't seems to be computed relatively to the die area. I mean by that that I would expect a placement density of 0.1 to be as sparse as possible on the die, but if the die is too small then the placement density also need to be reduced so I visualize some cells to be "outside" the die which OpenROAD don't like. 

But again, how does the target density argument is influencing the placement tool ? The density is computed with regard to what ?

**Comment by maliberty** [comment]
The layout is divided into bins of fixed area.  For each bin the density is the area of cells in the bin divided by the bin area.  Cells that want to cluster tightly together will be limited by the target density.  It has nothing to do with overall design area as you note.

**Reply by lustefan97** [reply]
Thanks for the answer ! What is the link between placement density and the die size then ? Why does OpenROAD ask to increase the placement density whenever the die size is deemed too small ? Or really, what does it mean to be "limited" by the placement density, what would it mean to be limited by a placement density of 0.1 vs 0.5 vs 1.0 ?

**Reply by maliberty** [reply]
The placement density implies a certain minimum area in which all the cells can be placed.  If that is larger than the available area then it is impossible to place at that density.  Increasing the density will reduce the required area so that it becomes possible.  The other choice would be to increase the total area.

**Reply by TanjIsGray** [reply]
But, why are we supposed to know that?  The layout density is whatever the tool can achieve, and we don't know that until the tool tries.  Now, it might make sense to say have a maximum density for power dissipation reasons, but even that would make more sense to say directly (like a target of no more than 2mw/mm2 which we could adjust).

It is especially puzzling when the run ends and tells us "65% does not work, 76% is estimated to be needed, try running again with higher density".  Well, duh.  If it can do that, why does it not just go ahead and do it?  It might make more sense to have a progress rule like "do the best you can, and stop when you are getting less than X% improvement per iteration" where X is a number with a default we can override.

OpenRoad has a goal to be automatic.  This is one parameter which is pretty strange to expect humans to provide, if the layout is supposed to be automatic.  We just end up making guesses.

**Reply by maliberty** [reply]
I agree it is an area for better automation (as is cell padding).
